### ISSUE TITLE
Terraform boats show "terraform" as the mod when viewed using the Jade mod

### ISSUE THREAD

**Issue by ghost** [TerraformersMC/Terraform#99]
I have This Problem

When I use a mod like the Jade mod, this mod puts the name of the thing referred to as shown in the picture (the white word is the name in the lang file and the blue word is the name of the mod itself, the name of my actual mod BetterBiomes, but the same name does not appear, rather it appears as terraform. Please solve For this problem.......

![2024-12-25_21 41 46](https://github.com/user-attachments/assets/4d0d7fe5-c373-4def-80f2-30c959a59e59)
![2024-12-25_21 48 20](https://github.com/user-attachments/assets/bae5b8e0-b896-4a69-a623-636dd2ee93b8)

**Comment by gniftygnome** [comment]
What version of Minecraft is this?  Also, please show your boat registration code; it sounds like it is being registered as something like `terraform:boat` which should not be the case.

**Comment by ghost** [comment]
I'm Using Minecraft 1.21  
Note : "Boat Word in Photo" is Name On Lang File , The Problem is from The Mod Name
This Code I'm Using
```

    public static final Item DRIFTWOOD_BOAT = TerraformBoatItemHelper.registerBoatItem(ModBoats.DRIFTWOOD_BOAT_ID, ModBoats.DRIFTWOOD_BOAT_KEY, false);
    public static final Item DRIFTWOOD_CHEST_BOAT = TerraformBoatItemHelper.registerBoatItem(ModBoats.DRIFTWOOD_CHEST_BOAT_ID, ModBoats.DRIFTWOOD_BOAT_KEY, true);

    public static final Identifier DRIFTWOOD_BOAT_ID = Identifier.of(BetterBiomes.MOD_ID, "driftwood_boat");
    public static final Identifier DRIFTWOOD_CHEST_BOAT_ID =  Identifier.of(BetterBiomes.MOD_ID, "driftwood_chest_boat");

    public static final RegistryKey<TerraformBoatType> DRIFTWOOD_BOAT_KEY = TerraformBoatTypeRegistry.createKey(DRIFTWOOD_BOAT_ID);

    public static void registerBoats() {
        TerraformBoatType driftwoodBoat = new TerraformBoatType.Builder()
                .item(ModItems.DRIFTWOOD_BOAT)
                .chestItem(ModItems.DRIFTWOOD_CHEST_BOAT)
                .planks(ModBlocks.DRIFTWOOD_PLANKS.asItem())
                .build();

        Registry.register(TerraformBoatTypeRegistry.INSTANCE, DRIFTWOOD_BOAT_KEY, driftwoodBoat);
```

**Comment by ghost** [comment]
look at TerraformBoatTypeRegistry Class ,
I saw it and I found

```
public class TerraformBoatTypeRegistry {
    private static final Identifier REGISTRY_ID = Identifier.of("terraform", "boat");
    private static final RegistryKey<Registry<TerraformBoatType>> REGISTRY_KEY;
    public static final Registry<TerraformBoatType> INSTANCE;

    public TerraformBoatTypeRegistry() {
    }

    public static RegistryKey<TerraformBoatType> createKey(Identifier id) {
        return RegistryKey.of(REGISTRY_KEY, id);
    }

    static {
        REGISTRY_KEY = RegistryKey.ofRegistry(REGISTRY_ID);
        INSTANCE = FabricRegistryBuilder.createSimple(REGISTRY_KEY).buildAndRegister();
    }
}

```

**Comment by gniftygnome** [comment]
> look at TerraformBoatTypeRegistry Class , I saw it and I found

That's the registery's ID, not any boat's ID.  The registry is indeed `terraform:boat`.

>  This Code I'm Using

Static constructors are essentially just static functions and are ordered (just like normal static functions).  Try moving the calls to `TerraformBoatItemHelper.registerBoatItem()` down below the init of DRIFTWOOD_BOAT_KEY.

**Comment by HealerPlusMC** [comment]
Can you help me by providing the correct code?

**Comment by gniftygnome** [comment]
Something like this is what I am suggesting.

```java
    public static final Identifier DRIFTWOOD_BOAT_ID = Identifier.of(BetterBiomes.MOD_ID, "driftwood_boat");
    public static final Identifier DRIFTWOOD_CHEST_BOAT_ID =  Identifier.of(BetterBiomes.MOD_ID, "driftwood_chest_boat");

    public static final RegistryKey<TerraformBoatType> DRIFTWOOD_BOAT_KEY = TerraformBoatTypeRegistry.createKey(DRIFTWOOD_BOAT_ID);

    public static final Item DRIFTWOOD_BOAT = TerraformBoatItemHelper.registerBoatItem(ModBoats.DRIFTWOOD_BOAT_ID, ModBoats.DRIFTWOOD_BOAT_KEY, false);
    public static final Item DRIFTWOOD_CHEST_BOAT = TerraformBoatItemHelper.registerBoatItem(ModBoats.DRIFTWOOD_CHEST_BOAT_ID, ModBoats.DRIFTWOOD_BOAT_KEY, true);
```

**Comment by HealerPlusMC** [comment]
it still **The Same Code**

**Comment by gniftygnome** [comment]
I'm sorry but what do you mean by that?  Are you saying you tried it and the result is the same as before?

**Comment by HealerPlusMC** [comment]
Yes, I Tried That

**Comment by gniftygnome** [comment]
While I can see why the code you pasted would fail, I do not see any significant difference between the code I pasted and the code which is known to be working in our mods.  I assume the function `registerBoats` is actually being called during mod init?  Also, please make sure you are calling the client side initializer in the client mod init.  Aside from checking those things I am not sure what might be going wrong.

**Comment by HealerPlusMC** [comment]
Look , Traverse Mod Has The Same Problem
![2024-12-29_11 34 05](https://github.com/user-attachments/assets/e86146be-79aa-41c7-8c8f-a6a20db4eb00)
![2024-12-29_11 34 08](https://github.com/user-attachments/assets/65bdb6a5-57d8-43d4-b415-5af7450c5747)

**Comment by HealerPlusMC** [comment]
The registerBoats function is indeed being called during the mod initialization.
The client-side initializer is properly invoked in the client mod initialization phase.


I think there is an incorrect register for any boat being registered

**Comment by gniftygnome** [comment]
![2024-12-29_01 59 49](https://github.com/user-attachments/assets/eb7077ea-8a80-4543-afeb-2cc09d9be4bd)

My best guess is maybe there's something odd about how Jade decides what to display there.  This image is using WTHIT.  I've never used Jade before so I'm not too familiar with its quirks.

**Comment by HealerPlusMC** [comment]
How !!

**Comment by gniftygnome** [comment]
FWIW there's a good chance this is a non-issue in 1.21.2 and above.  Because of major vanilla changes, the boat code was completely rewritten to use real boat entities instead of impersonating the oak boats or bamboo rafts.

**Comment by gniftygnome** [comment]
![2024-12-29_02 20 24](https://github.com/user-attachments/assets/0cef3355-0b96-4cd6-a0b5-13662ace6f9e)

Confirmed.  I reproduced this with Jade in 1.21.1.  However, the above is how it will look with Jade in 1.21.4.  If I find some time I may look into whether we can somehow accommodate their approach in 1.21.1 but I suspect they are just taking a literal ID namespace from somewhere, which probably isn't easy to accommodate.

**Comment by HealerPlusMC** [comment]
hmmmm. I think That Too

**Comment by HealerPlusMC** [comment]
I will Try To Find The Solve :(

**Comment by HealerPlusMC** [comment]
Wait a minute look at this
![2024-12-29_17 05 53](https://github.com/user-attachments/assets/919a2fbb-86e9-4477-997c-7aa34511ba01)
![2024-12-29_17 05 58](https://github.com/user-attachments/assets/243a0070-ca29-453a-9ef6-b9d9585bc64d)
![2024-12-29_17 06 01](https://github.com/user-attachments/assets/3729ffb3-7c42-4c12-bc94-6ae620d2ee07)

**Comment by HealerPlusMC** [comment]
Half the problem is **Terraform** itself , and the other foolish half of **Jade**

**Comment by gniftygnome** [comment]
Yes, I was testing in 1.21.1.

At the risk of rehashing something which does not have much point (since this issue no longer exists since 1.21.2), Terraform until 1.21.2 created its own boat entities which impersonated oak boats or bamboo rafts but rendered as whatever the Terraform boat actually was.  This process worked just like how vanilla boats were implemented.  These entities had entity IDs like `terraform:boat` or `terraform:chest_raft` in exactly the same way that Minecraft's own boats had entity IDs like `minecraft:boat` or `minecraft:chest_raft`.  When something wants to know about a boat entity prior to 1.21.2, it calls `boatEntity.getVariant()` (which returns the `BoatEntity.Type`) and the boat entity type will have the real boat variant ID of the boat entity.  What Jade must be doing is just using the boatEntity's entity ID (remember this is something like `minecraft:chest_boat` or `terraform:raft`) namespace and displaying any translation available for that value, or the raw namespace if none.  What it *should* be doing prior to 1.21.2 is calling the getVariant() method and displaying the ID namespace of the returned boat variant.

It is easy to imagine other boat implementations aside from Terraform might be using vanilla boat entities, in which case their boats would be attributed by Jade to `Minecraft` which is perhaps even more off-track than `terraform`...

In a strict sense, the value Jade is displaying is arguably "correct" in the case of Terraform boats, since it is showing the mod which is most directly responsible for adding the boat.  However, from the user's perspective, what Jade is doing is wrong.  In any case, it is their decision.  WTHIT makes the more user-friendly decision and displays the values from the boat's variant.

**Comment by HealerPlusMC** [comment]
Can You fix this error for version 1.21?

**Comment by HealerPlusMC** [comment]
I'm working on a big project
I am still a beginner and I want to make no **mistakes**

**Comment by HealerPlusMC** [comment]
I know this might be difficult, but try..

**Comment by HealerPlusMC** [comment]
Hello, Where Are you "gniftygnome"

**Comment by gniftygnome** [comment]
I do not think we can "fix" this prior to 1.21.2.  The changes in 1.21.2 can't be backported because they were based on huge changes to the vanilla boat implementation.  Considering the fact this is how Terraform boat API has worked for years and we have never had any complaint until now, I also do not think this is urgent relative to other ongoing work.  Furthermore, I do not have the time to go on a futile goose chase looking for a fix when Jade is very likely just using the entity ID namespace which we cannot change.  This should be fairly easy for Jade to fix but it would be very difficult for us to fix.

**Comment by gniftygnome** [comment]
Or you could switch to WTHIT and the problem would magically go away, which seems a fine work-around...

**Comment by HealerPlusMC** [comment]
If you say you fixed the problem This is Good
Because you rarely find anyone using Jade Mod
Except for those who use Better MC Mod Pack

Close This Issues

**Comment by gniftygnome** [comment]
It is fixed, from 1.21.2, yes.  I've looked into this further and there is no good way for us to fix it in 1.21.1, although Jade could do so by switching to use the item pick stack for the boat which it turns out is what WTHIT is doing.

I will resolve fixed, since it is in newer versions.  Good luck with your project!
